File: iosApp/Sources/Pages/OfflinePage.swift
import IterCore
import SwiftUI

/// Offline areas: download the current viewport as a bundle, list and delete
/// installed areas.
struct OfflinePage: View {
  /// Supplies the map viewport as the wire bbox string, nil before readiness.
  let currentBBox: () -> String?

  @EnvironmentObject private var offline: OfflineModel

  var body: some View {
    List {
      Section {
        downloadRow
      }
      Section {
        if offline.areas.isEmpty {
          Text(Strings.offlineEmpty)
            .foregroundStyle(.secondary)
        } else {
          ForEach(offline.areas, id: \.id) { area in
            AreaRow(area: area) {
              offline.delete(areaId: area.id)
            }
          }
        }
      }
    }
    .navigationTitle(Strings.offlineTitle)
    .navigationBarTitleDisplayMode(.inline)
    .onAppear { offline.refresh() }
  }

  @ViewBuilder
  private var downloadRow: some View {
    switch offline.download {
    case .downloading(let read, let total):
      VStack(alignment: .leading, spacing: 8) {
        if let total, total > 0 {
          ProgressView(value: Double(read), total: Double(total))
        } else {
          ProgressView()
        }
        Text(Strings.offlineDownloading(Formatters.bytes(read)))
          .font(.footnote)
          .foregroundStyle(.secondary)
      }
    case .installing:
      VStack(alignment: .leading, spacing: 8) {
        ProgressView()
        Text(Strings.offlineInstalling)
          .font(.footnote)
          .foregroundStyle(.secondary)
      }
    case .failed(let code):
      VStack(alignment: .leading, spacing: 8) {
        Text(failureMessage(code))
          .font(.subheadline)
          .foregroundStyle(.red)
        // Retry the download rather than merely clearing the error.
        downloadButton(Strings.actionRetry)
      }
    case .idle:
      downloadButton(Strings.offlineDownloadCurrent)
    }
  }

  private func downloadButton(_ title: String) -> some View {
    Button {
      if let bbox = currentBBox() {
        offline.downloadArea(bbox: bbox)
      }
    } label: {
      Label(title, systemImage: "arrow.down.circle.fill")
    }
  }

  private func failureMessage(_ code: String?) -> String {
    if OfflineModel.isTooLarge(code) { return Strings.offlineErrorTooLarge }
    if OfflineModel.isInvalidArea(code) { return Strings.offlineErrorInvalidArea }
    if OfflineModel.isBusy(code) { return Strings.offlineErrorBusy }
    return Strings.errorGeneric
  }
}

private struct AreaRow: View {
  let area: OfflineArea
  let onDelete: () -> Void

  var body: some View {
    HStack {
      VStack(alignment: .leading, spacing: 2) {
        Text(label)
        Text("\(Strings.offlineAreaZoom(Int(area.manifest.maxzoom)))  ·  \(String(area.manifest.createdAt.prefix(10)))")
          .font(.footnote)
          .foregroundStyle(.secondary)
      }
      Spacer()
      Button(role: .destructive, action: onDelete) {
        Image(systemName: "trash")
      }
      .buttonStyle(.borderless)
      .accessibilityLabel(Strings.actionDelete)
    }
  }

  private var label: String {
    // Manifest bbox is the JSON 4-tuple [minLon, minLat, maxLon, maxLat].
    let bbox = area.manifest.bbox.map { $0.doubleValue }
    guard bbox.count == 4 else { return area.id }
    return String(format: "%.2f, %.2f → %.2f, %.2f", bbox[1], bbox[0], bbox[3], bbox[2])
  }
}

File: androidApp/src/main/kotlin/it/iterapp/app/offline/OfflinePage.kt
package it.iterapp.app.offline

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.rounded.CloudDownload
import androidx.compose.material.icons.rounded.Delete
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import it.iterapp.app.R
import it.iterapp.app.sheet.SheetPageHeader
import it.iterapp.core.offline.OfflineArea

/**
 * Offline areas: download the current viewport as a bundle, list and delete
 * installed areas. Errors map the contract's codes to friendly copy.
 */
@Composable
fun OfflinePage(
  viewModel: OfflineViewModel,
  currentViewportBBox: () -> String?,
  onBack: () -> Unit,
  modifier: Modifier = Modifier,
) {
  val areas by viewModel.areas.collectAsStateWithLifecycle()
  val download by viewModel.download.collectAsStateWithLifecycle()

  Column(modifier.fillMaxSize()) {
    SheetPageHeader(title = stringResource(R.string.offline_title), onBack = onBack)

    Column(Modifier.padding(horizontal = 16.dp)) {
      when (val d = download) {
        is DownloadState.Downloading -> {
          val progress = d.total?.let { total ->
            if (total > 0) d.bytesRead.toFloat() / total else null
          }
          if (progress != null) {
            LinearProgressIndicator(
              progress = { progress },
              modifier = Modifier.fillMaxWidth(),
            )
          } else {
            LinearProgressIndicator(modifier = Modifier.fillMaxWidth())
          }
          Text(
            text = stringResource(
              R.string.offline_downloading,
              formatBytes(d.bytesRead),
            ),
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier.padding(vertical = 8.dp),
          )
        }
        DownloadState.Installing -> {
          LinearProgressIndicator(modifier = Modifier.fillMaxWidth())
          Text(
            text = stringResource(R.string.offline_installing),
            style = MaterialTheme.typography.bodySmall,
            modifier = Modifier.padding(vertical = 8.dp),
          )
        }
        is DownloadState.Failed -> {
          // Error plus a retry button, so a failure is never a dead end.
          Text(
            text = stringResource(
              when {
                OfflineViewModel.isTooLarge(d.code) -> R.string.offline_error_too_large
                OfflineViewModel.isInvalidArea(d.code) -> R.string.offline_error_invalid_area
                OfflineViewModel.isBusy(d.code) -> R.string.offline_error_busy
                else -> R.string.error_generic
              },
            ),
            color = MaterialTheme.colorScheme.error,
            style = MaterialTheme.typography.bodyMedium,
            modifier = Modifier.padding(vertical = 8.dp),
          )
          DownloadButton(currentViewportBBox, viewModel, label = R.string.action_retry)
        }
        DownloadState.Idle -> DownloadButton(currentViewportBBox, viewModel)
      }
    }

    if (areas.isEmpty()) {
      Text(
        text = stringResource(R.string.offline_empty),
        color = MaterialTheme.colorScheme.onSurfaceVariant,
        style = MaterialTheme.typography.bodyMedium,
        modifier = Modifier.padding(horizontal = 24.dp, vertical = 20.dp),
      )
    } else {
      LazyColumn(Modifier.fillMaxSize().padding(top = 8.dp)) {
        items(areas, key = { it.id }) { area ->
          AreaRow(area = area, onDelete = { viewModel.delete(area.id) })
        }
      }
    }
  }
}

@Composable
private fun DownloadButton(
  currentViewportBBox: () -> String?,
  viewModel: OfflineViewModel,
  label: Int = R.string.offline_download_current,
) {
  Button(
    onClick = { currentViewportBBox()?.let { viewModel.downloadArea(it) } },
    shape = MaterialTheme.shapes.medium,
    modifier = Modifier.fillMaxWidth(),
  ) {
    Icon(
      Icons.Rounded.CloudDownload,
      contentDescription = null,
      modifier = Modifier.size(ButtonDefaults.IconSize),
    )
    Spacer(Modifier.width(ButtonDefaults.IconSpacing))
    Text(text = stringResource(label))
  }
}

@Composable
private fun AreaRow(area: OfflineArea, onDelete: () -> Unit) {
  Surface(color = androidx.compose.ui.graphics.Color.Transparent) {
    Row(
      verticalAlignment = Alignment.CenterVertically,
      modifier = Modifier
        .fillMaxWidth()
        .padding(horizontal = 16.dp, vertical = 6.dp),
    ) {
      Column(Modifier.weight(1f)) {
        Text(
          text = areaLabel(area),
          style = MaterialTheme.typography.bodyLarge,
        )
        Text(
          text = stringResource(R.string.offline_area_zoom, area.manifest.maxzoom) +
            "  ·  " + area.manifest.createdAt.take(10),
          style = MaterialTheme.typography.bodySmall,
          color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
      }
      IconButton(onClick = onDelete) {
        Icon(
          Icons.Rounded.Delete,
          contentDescription = stringResource(R.string.action_delete),
          tint = MaterialTheme.colorScheme.error,
        )
      }
    }
  }
}

private fun areaLabel(area: OfflineArea): String {
  val bbox = area.manifest.bbox
  return if (bbox.size == 4) {
    "%.2f, %.2f → %.2f, %.2f".format(bbox[1], bbox[0], bbox[3], bbox[2])
  } else {
    area.id
  }
}

private fun formatBytes(bytes: Long): String = when {
  bytes < 1024 * 1024 -> "${bytes / 1024} KB"
  else -> "%.1f MB".format(bytes / (1024.0 * 1024.0))
}
